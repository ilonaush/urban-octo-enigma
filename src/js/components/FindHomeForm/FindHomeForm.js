import React, {Component} from 'react';
import PropTypes from 'prop-types';
import "./FindHomeForm.styl";
import Input from "../Input/Input";

export class FindHomeForm extends Component {

    constructor(props) {
        super(props);
        this.state = {
            catID: '',
        };
    }


    handleSelect = ({target = null}) => {
        this.setState({
            catID: target.value
        })
    };

    render() {
        const {onSubmit, cats} = this.props;
        const {catID} = this.state;
        return (
            <form className='fireWorkerForm' onSubmit={(e) => onSubmit(e, catID)}>
                <select onChange={this.handleSelect}>
                    <option value=" ">Choose the meowy</option>
                    {cats.map((cat) =>
                        <option key={cat.id} value={cat.id} >{cat.name}, {cat.age}</option>
                    )}
                </select>
                <Input placeholder='Location of future home'/>
                <textarea placeholder='Describe the family please'/>
                <button id='submit-btn' type='submit'>Find home for the cat</button>
            </form>
        );
    }
}

FindHomeForm.propTypes = {
    cats: PropTypes.array,
    onSubmit: PropTypes.func
};

FindHomeForm.defaultProps = {
    cats: [],
};

export default FindHomeForm;
