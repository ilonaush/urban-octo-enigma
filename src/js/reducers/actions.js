import RequestService from "../services/RequestService";
import {ACTIONS_TYPES, LOADING_TYPES, REQUEST_PATHS} from "../constants";

/**
 * sets loading state and loader type
 * @param type
 * @param dispatch
 * @returns {*}
 */
function setLoading(type, dispatch) {
    return dispatch({
        type: ACTIONS_TYPES.SET_LOADING,
        payload: {loading: true, loadingType: type}
    });
}

/**
 * dispatches empty array
 * @param dispatch
 * @param type
 */
function handleError(dispatch, type) {
    dispatch({
        type: type,
        payload: { cats: [], loading: false }
    })
}

export default {
    getCats: function() {
        return  async (dispatch) => {

            setLoading(null, dispatch);

            try {
                const {data: cats} = await RequestService.get(REQUEST_PATHS.GET_CATS);
                dispatch({
                    type: ACTIONS_TYPES.GET_CATS,
                    payload: { cats, loading: false }
                })
            }
            catch (e) {
              handleError(dispatch, ACTIONS_TYPES.GET_CATS)
            }
        };
    },
    addCat: function (cat) {
        return async (dispatch) => {

            setLoading(null, dispatch);

            try {
                const {data: {cats, status}} = await RequestService.post(REQUEST_PATHS.ADD_CAT, cat);
                if (!status) {
                    handleError(dispatch, ACTIONS_TYPES.ADD_CAT)
                }
                dispatch({
                    type: ACTIONS_TYPES.ADD_CAT,
                    payload: {cats, loading: false}
                });
            }
            catch (e) {
                handleError(dispatch, ACTIONS_TYPES.ADD_CAT)
            }
        }
    },
    feedCat(catInfo) {
        return async (dispatch) =>  {

            setLoading(LOADING_TYPES.FEED_CAT, dispatch);

            try {
                const {data: {cats, status}} = await RequestService.patch(REQUEST_PATHS.FEED_CAT, catInfo);
                if (!status) {
                    handleError(dispatch, ACTIONS_TYPES.FEED_CAT)
                }
                dispatch({
                    type: ACTIONS_TYPES.FEED_CAT,
                    payload: {cats, loading: false}
                });
            }
            catch (e) {
                handleError(dispatch, ACTIONS_TYPES.FEED_CAT)

            }
    }},
    hugCat(catInfo) {
        return async (dispatch) =>  {

            setLoading(LOADING_TYPES.HUG_CAT, dispatch);

            try {
                const {data: {cats, status}} = await RequestService.patch(REQUEST_PATHS.HUG_CAT, catInfo);
                if (!status) {
                    handleError(dispatch, ACTIONS_TYPES.HUG_CAT)
                }
                dispatch({
                    type: ACTIONS_TYPES.HUG_CAT,
                    payload: {cats, loading: false}
                });
            }
            catch (e) {
                handleError(dispatch, ACTIONS_TYPES.HUG_CAT)
            }
    }},
    washCat(catInfo) {
        return async (dispatch) =>  {

            setLoading(LOADING_TYPES.WASH_CAT, dispatch);

            try {
                const {data: {cats, status}} = await RequestService.patch(REQUEST_PATHS.WASH_CAT, catInfo);
                if (!status) {
                    handleError(dispatch, ACTIONS_TYPES.WASH_CAT)
                }
                dispatch({
                    type: ACTIONS_TYPES.WASH_CAT,
                    payload: {cats, loading: false}
                });
            }
            catch (e) {
                handleError(dispatch, ACTIONS_TYPES.WASH_CAT)
            }
    }},
    findHome: function (cat) {
        return async (dispatch) =>  {

            setLoading(null, dispatch);

            try {
                const variables = {
                    ...cat
                };

                const {data: {cats, status}} = await RequestService.patch(REQUEST_PATHS.ISSUE_CAT, variables);
                if (!status) {
                    handleError(dispatch, ACTIONS_TYPES.ISSUE_CAT)
                }
                dispatch({
                    type: ACTIONS_TYPES.ISSUE_CAT,
                    payload: {cats, loading: false}
                });
            }
            catch (e) {
                handleError(dispatch, ACTIONS_TYPES.ISSUE_CAT)
            }
        }
    }
};
