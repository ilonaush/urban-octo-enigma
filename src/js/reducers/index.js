import { createStore, applyMiddleware } from 'redux';
import reducer   from './reducers';
import  { saveStore, loadStore } from '../services/LocalStorageService';
import thunk from 'redux-thunk'
import { createLogger } from 'redux-logger'

const loggerMiddleware = createLogger();

// const predefinedStore = loadStore();

const store = createStore(
    reducer,
    // predefinedStore,
    applyMiddleware(thunk, loggerMiddleware)
);

// store.subscribe(() => {
//     saveStore(store.getState())
// });

export {
    store,
    // predefinedStore
};
