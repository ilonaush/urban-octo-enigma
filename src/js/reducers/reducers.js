import {ACTIONS_TYPES} from "../constants";

const initialState = {
  loading: false,
};

let reducer = (state = initialState, action) => {
  switch (action.type) {
    case ACTIONS_TYPES.GET_CATS:
      return {
        ...state,
        cats: action.payload.cats,
        loading: action.payload.loading,
        loadingType: action.payload.loadingType
      };
    case ACTIONS_TYPES.ADD_CAT:
      return {
        ...state,
        cats: action.payload.cats,
        loading: action.payload.loading,
        loadingType: action.payload.loadingType
      };
    case ACTIONS_TYPES.FEED_CAT:
      return {
        ...state,
        cats: action.payload.cats,
        loading: action.payload.loading,
        loadingType: action.payload.loadingType
      };
    case ACTIONS_TYPES.HUG_CAT:
      return {
        ...state,
        cats: action.payload.cats,
        loading: action.payload.loading,
        loadingType: action.payload.loadingType
      };
    case ACTIONS_TYPES.WASH_CAT:
      return {
        ...state,
        cats: action.payload.cats,
        loading: action.payload.loading,
        loadingType: action.payload.loadingType
      };
    case ACTIONS_TYPES.ISSUE_CAT:
      return {
        ...state,
        cats: action.payload.cats,
        loading: action.payload.loading,
        loadingType: action.payload.loadingType
      };
    case ACTIONS_TYPES.SET_LOADING: {
      return {
        ...state,
        loading: action.payload.loading,
        loadingType: action.payload.loadingType
      }
    }
    default:
      return state;
  }
};

export default reducer;
