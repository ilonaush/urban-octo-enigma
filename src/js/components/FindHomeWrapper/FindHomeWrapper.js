import React, {Component, Fragment} from 'react';
import actions from "../../reducers/actions";
import {connect} from "react-redux";
import {FindHomeForm} from "../FindHomeForm/FindHomeForm";
import PropTypes from 'prop-types';
import moment from "moment";
import {store} from "../../reducers";


export class FindHomeWrapper extends Component {

  constructor(props) {
    super(props);
  }

  handleSubmit = (e, cat) => {
    const variables = {
      ...cat,
      date: moment()
    };
    e.preventDefault();
    this.props.findHome(variables)
        .then(() => this.props.history.push('/'))
        .catch(e => console.log(e));
  };

  render() {
    return (
        <div>
          <FindHomeForm onSubmit={this.handleSubmit} cats={this.props.cats}/>
          <img className='cat-bg' src="/images/backgrounds/cat-bg-1.png" alt=""/>
        </div>
    );
  }
}

FindHomeWrapper.propTypes = {
  cats: PropTypes.array,
  findHome: PropTypes.func
};

export default connect((state) => ({cats: state.cats}),
    (dispatch) => ({findHome: (catID) => dispatch(actions.findHome(catID))}))(FindHomeWrapper);
