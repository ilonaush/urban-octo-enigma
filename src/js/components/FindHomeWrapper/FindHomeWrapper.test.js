import React from 'react';
import {FindHomeWrapper}  from './FindHomeWrapper';
import {Provider} from "react-redux";
import {store} from "../../reducers/index";
import {mount, shallow} from "enzyme/build";
import renderer from "react-test-renderer";
import FindHomeForm from "../FindHomeForm/FindHomeForm";



describe('FindHomeWrapper', () => {
    it('renders correctly', () => {
        const FireWorkerWrapperComponent = renderer.create(<FindHomeWrapper/>);
        expect((FireWorkerWrapperComponent).toJSON()).toMatchSnapshot();
    });

    it('calls submit function  when  the form is submitted', async () => {
        const historyMock = { push: jest.fn() };
        const mockFindHomeAction = jest.fn().mockResolvedValue();
        const props = {
            findHome: mockFindHomeAction,
            history: historyMock
        };
        const FireWorkerWrapperComponent = mount(
            <Provider store={store}>
                <FindHomeWrapper {...props}/>
            </Provider>
        );

        const fireWorkerFormComponent = FireWorkerWrapperComponent.find(FindHomeForm);
        expect(mockFindHomeAction.mock.calls.length).toBe(0);
        fireWorkerFormComponent.simulate('submit');
        expect(mockFindHomeAction.mock.calls.length).toBe(1)
    });


});

