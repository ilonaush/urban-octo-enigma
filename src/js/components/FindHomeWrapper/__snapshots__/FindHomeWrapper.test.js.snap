// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FindHomeWrapper renders correctly 1`] = `
<div>
  <form
    className="fireWorkerForm"
    onSubmit={[Function]}
  >
    <select
      name="id"
      onChange={[Function]}
      required={true}
    >
      <option
        value=" "
      >
        Choose the meowy
      </option>
    </select>
    <input
      name="address"
      onChange={[Function]}
      placeholder="Location of future home"
      required={true}
      value=""
    />
    <textarea
      name="family"
      onChange={[Function]}
      placeholder="Describe the family please"
      required={true}
    />
    <button
      id="submit-btn"
      type="submit"
    >
      Find home for the cat
    </button>
  </form>
  <img
    alt=""
    className="cat-bg"
    src="/images/backgrounds/cat-bg-1.png"
  />
</div>
`;
